`jax.nn.standardize` produces different outputs in eager vs. outer-JIT context due to numerically unstable variance formula

### Description

`jax.nn.standardize` is decorated with `@jax.jit` internally and uses the
König–Huygens variance formula `var = mean(x²) − mean(x)²`. When a model
containing `jax.nn.standardize` is compiled with an **outer** `jax.jit`, XLA
fuses the standardize computation with upstream operations and applies different
floating-point optimizations than when standardize runs as a standalone pre-JIT
call on already-computed eager values. The numerically sensitive subtraction in
the variance formula amplifies these tiny differences, producing a max output
discrepancy of ~0.003 — well above the oracle threshold of 1e-3 — even though
the inputs to standardize are bit-for-bit identical.

---

## Possible Root cause

`jax.nn.standardize` computes variance as:

```python
variance = jnp.mean(jnp.square(x), axis, keepdims=True) - jnp.square(mean)
```

This is the König–Huygens (computational) formula. It is numerically unstable
when the variance is small relative to the mean-squared term, because it
subtracts two nearly equal large numbers. The JAX source even acknowledges
this: *"this definition is traditionally seen as less accurate than
`jnp.var`'s `mean((x - mean(x))**2)`"*.

The instability is exposed here because:
1. `jax.nn.standardize` is `@jit`-decorated, so it compiles as an isolated XLA
   program when called in a non-JIT context.
2. When the user wraps their model in an outer `jax.jit`, the inner `@jit` is
   a no-op — XLA re-traces the whole model and compiles one fused XLA program.
3. In the fused program, XLA sees the gelu/silu/reduce_window outputs as
   intermediate values and may reorder floating-point reductions differently
   than in the standalone compile, yielding a slightly different variance.
4. With `var ≈ 2.3e-5` and `epsilon = 1e-5`, the denominator
   `sqrt(var + epsilon) ≈ 5.7e-3` is small, amplifying the variance error by
   ~125× in the final normalized output.

**Switching to the two-pass (stable) formula reduces max_diff from 0.003 to
0.000024**, confirming the variance formula is the primary driver.

---

## Minimal reproducer

```python
import jax
import jax.numpy as jnp

key = jax.random.PRNGKey(121235)
key, k = jax.random.split(key)
x = jax.random.normal(k, (15, 14, 13, 4))

def model(x):
    t1 = jax.lax.reduce_window(x, -jnp.inf, jax.lax.max, (1, 1, 2, 2), (1, 1, 2, 2), 'VALID')
    t2 = jax.nn.silu(t1)
    t3 = jax.nn.gelu(t2)
    return jax.nn.standardize(t3, axis=-1)

eager_out = model(x)          # standardize runs as its own @jit
jit_out   = jax.jit(model)(x) # standardize fused into outer JIT

max_diff = float(jnp.abs(eager_out - jit_out).max())
print(f"max_diff = {max_diff:.6f}")  # 0.003024
assert max_diff < 1e-3, f"eager/jit mismatch: {max_diff}"
```

Expected: assertion passes (outputs agree within 1e-3).  
Actual: `AssertionError: eager/jit mismatch: 0.003024`

The mismatch occurs on 78/100 random seeds with this input shape.


### System info (python version, jaxlib version, accelerator, etc.)

```
jax:    0.6.2
jaxlib: 0.6.2
numpy:  2.2.6
python: 3.10.12 (main, Mar  3 2026, 11:56:32) [GCC 11.4.0]
device info: cpu-1, 1 local devices"
process_count: 1
platform: uname_result(system='Linux', node='node1', release='6.8.0-94-generic', version='#96~22.04.1-Ubuntu SMP PREEMPT_DYNAMIC Fri Jan 16 13:19:05 UTC 2', machine='x86_64')
```

--- a/jax/_src/nn/functions.py
+++ b/jax/_src/nn/functions.py
@@ -645,13 +645,15 @@ def _softmax_deprecated(
   return result
 
 
-@api.jit(static_argnames=("axis",))
+@api.jit(static_argnames=("axis", "algorithm"))
 def standardize(x: ArrayLike,
                 axis: Axis = -1,
                 mean: ArrayLike | None = None,
                 variance: ArrayLike | None = None,
                 epsilon: ArrayLike = 1e-5,
-                where: ArrayLike | None = None) -> Array:
+                where: ArrayLike | None = None,
+                *,
+                algorithm: str = "fast") -> Array:
   r"""Standardizes input to zero mean and unit variance.
 
   The standardization is given by:
@@ -675,6 +677,11 @@ def standardize(x: ArrayLike,
       to ``1E-5``.
     where: optional boolean mask specifying which elements to use when computing
       the mean and variance.
+    algorithm: variance computation algorithm. ``"fast"`` uses ``mean(x^2) - mean(x)^2``
+      which may be faster but can suffer from catastrophic cancellation and produce
+      different results in eager vs JIT contexts. ``"stable"`` uses the two-pass
+      formula ``mean((x - mean(x))^2)`` which is numerically stable. Default is
+      ``"fast"`` for backward compatibility.
 
   Returns:
     An array of the same shape as ``x`` containing the standardized input.
@@ -684,17 +691,23 @@ def standardize(x: ArrayLike,
   if mean is None:
     mean = jnp.mean(x, axis, keepdims=True, where=where)
   if variance is None:
-    # this definition is traditionally seen as less accurate than jnp.var's
-    # mean((x - mean(x))**2) but may be faster and even, given typical
-    # activation distributions and low-precision arithmetic, more accurate
-    # when used in neural network normalization layers
-    variance = jnp.mean(
-        jnp.square(x), axis, keepdims=True, where=where) - jnp.square(mean)
-    # Because we're using a less accurate variance definition, it may return
-    # negative values. This is problematic for the rsqrt, so we clip to 0.
-    # Note that this clipping only matters when the variance is vanishingly
-    # small compared to the mean of x, so the gradient should be unaffected.
-    variance = jnp.clip(variance, 0)
+    if algorithm == "stable":
+      variance = jnp.mean(
+          jnp.square(jnp.subtract(x, mean)), axis, keepdims=True, where=where)
+    elif algorithm == "fast":
+      # This definition is traditionally seen as less accurate than the
+      # two-pass mean((x - mean(x))**2) but may be faster and even, given
+      # typical activation distributions and low-precision arithmetic, more
+      # accurate when used in neural network normalization layers.
+      variance = jnp.mean(
+          jnp.square(x), axis, keepdims=True, where=where) - jnp.square(mean)
+      # Because we're using a less accurate variance definition, it may
+      # return negative values. This is problematic for the rsqrt, so we
+      # clip to 0.
+      variance = jnp.clip(variance, 0)
+    else:
+      raise ValueError(
+          f"Unknown algorithm '{algorithm}'. Expected 'fast' or 'stable'.")
   return jnp.subtract(x, mean) * lax.rsqrt(variance + epsilon)
 
 # TODO(slebedev): Change the type of `x` to `ArrayLike`.

--- a/tests/nn_test.py
+++ b/tests/nn_test.py
@@ -664,6 +664,20 @@ class NNFunctionsTest(jtu.JaxTestCase):
     result = jax.nn.standardize(x)
     self.assertFalse(jnp.any(jnp.isnan(result)))
 
+  def testStandardizeStableAlgorithm(self):
+    # Regression test for https://github.com/jax-ml/jax/issues/37281
+    # The stable algorithm should give consistent results regardless of
+    # compilation context.
+    x = jnp.array([1e6, 1e6 + 1, 1e6 + 2], dtype=jnp.float32)
+    result_eager = nn.standardize(x, algorithm="stable")
+    result_jit = jax.jit(lambda x: nn.standardize(x, algorithm="stable"))(x)
+    self.assertAllClose(result_eager, result_jit, atol=1e-5)
+    # Should be close to [-1, 0, 1] * scale
+    self.assertFalse(jnp.any(jnp.isnan(result_eager)))
+    # Fast algorithm should also work (backward compat)
+    result_fast = nn.standardize(x, algorithm="fast")
+    self.assertFalse(jnp.any(jnp.isnan(result_fast)))
+
   def testOneHot(self):
     actual = nn.one_hot(jnp.array([0, 1, 2]), 3)
     expected = jnp.array([[1., 0., 0.],

